C/STrace/EtwLogger.h

```diff
@@ -0,0 +1,314 @@
+
+	return eventMetadataDesc;
+}
+
+// final recursion case, string-specific overload
+NTSTATUS CreateTracePropertyRecursive(OUT PEVENT_DATA_DESCRIPTOR propertyDataDescriptors, const char* fieldName, int fieldType, const char* fieldValue)
+{
+	// fieldName and fieldType are used in the event metadata descriptor, not here.
+	UNREFERENCED_PARAMETER(fieldName);
+	UNREFERENCED_PARAMETER(fieldType);
+
+	// Copy the input value to its own space.
+	const auto newSpace = (char*)ExAllocatePoolWithTag(NonPagedPoolNx, strlen(fieldValue) + 1, 'wteE');
+	if (newSpace == NULL)
+	{
+		return STATUS_UNSUCCESSFUL;
+	}
+	strcpy(newSpace, fieldValue);
+
+	// Create the event data descriptor pointing to the value of the field.
+	EventDataDescCreate(OUT propertyDataDescriptors, newSpace, (ULONG)(strlen(fieldValue) + 1));
+
+	return STATUS_SUCCESS;
+}
+
+// final recursion case
+template<typename FieldValue>
+NTSTATUS CreateTracePropertyRecursive(OUT PEVENT_DATA_DESCRIPTOR propertyDataDescriptors, const char* fieldName, int fieldType, FieldValue& fieldValue)
+{
+	// fieldName and fieldType are used in the event metadata descriptor, not here.
+	UNREFERENCED_PARAMETER(fieldName);
+	UNREFERENCED_PARAMETER(fieldType);
+
+	// Copy the input value to its own space.
+	const auto newSpace = ExAllocatePoolWithTag(NonPagedPoolNx, sizeof(FieldValue), 'wteE');
+	if (newSpace == NULL)
+	{
+		return STATUS_UNSUCCESSFUL;
+	}
+	memcpy(newSpace, &fieldValue, sizeof(FieldValue));
+
+	// Create the event data descriptor pointing to the value of the field.
+	EventDataDescCreate(OUT propertyDataDescriptors, newSpace, sizeof(FieldValue));
+
+	return STATUS_SUCCESS;
+}
+
+// string-specific overload
+template<typename... Rest>
+NTSTATUS CreateTracePropertyRecursive(OUT PEVENT_DATA_DESCRIPTOR propertyDataDescriptors, const char* fieldName, int fieldType, const char* fieldValue, Rest... rest)
+{
+	// fieldName and fieldType are used in the event metadata descriptor, not here.
+	UNREFERENCED_PARAMETER(fieldName);
+	UNREFERENCED_PARAMETER(fieldType);
+
+	// Copy the input value to its own space.
+	const auto newSpace = (char*)ExAllocatePoolWithTag(NonPagedPoolNx, strlen(fieldValue) + 1, 'wteE');
+	if (newSpace == NULL)
+	{
+		return STATUS_UNSUCCESSFUL;
+	}
+	strcpy(newSpace, fieldValue);
+
+	// Create the event data descriptor pointing to the value of the field.
+	EventDataDescCreate(OUT propertyDataDescriptors, newSpace, (ULONG)(strlen(fieldValue) + 1));
+
+	// Create the next descriptor with the next field value.
+	return CreateTracePropertyRecursive(++propertyDataDescriptors, rest...);
+}
+
+template<typename FieldValue, typename... Rest>
+NTSTATUS CreateTracePropertyRecursive(OUT PEVENT_DATA_DESCRIPTOR propertyDataDescriptors, const char* fieldName, int fieldType, FieldValue& fieldValue, Rest... rest)
+{
+	// fieldName and fieldType are used in the event metadata descriptor, not here.
+	UNREFERENCED_PARAMETER(fieldName);
```

Comment:
This is a result of me not knowing too much about templates - I think if we want to go through each of the arguments as a triplet (name, type, value), then they need to be defined? Otherwise the first function call would use the name as the value, the second the type as the value, and so on? But then the function doesn't actually need them.

This should be fixed with the C-style varargs.

Reply:
Oh I see why this was done now, yea this isn't the easiest thing to do in C++. You would probably have to convert the arguments into a tuple first or something like that. I would need to play with it, but since we're converting to printf style we can ignore this. If you need to keep these around that's fine for now.